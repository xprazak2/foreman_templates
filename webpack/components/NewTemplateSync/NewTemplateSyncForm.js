import React from 'react';
import { reduxForm, formValueSelector, change } from 'redux-form';
import { connect } from 'react-redux';
import reduce from 'ramda/src/reduce';
import concat from 'ramda/src/concat';
import curry from 'ramda/src/curry';

import Form from 'foremanReact/components/common/forms/Form';
import RadioButtonGroup from 'foremanReact/components/common/forms/RadioButtonGroup';
import * as FormActions from 'foremanReact/redux/actions/common/forms';
import TextField from 'foremanReact/components/common/forms/TextField';

import SyncSettingsFields from './SyncSettingFields';
import Title from '../layout/Title';

const formName = 'newTemplateSync';

const redirect = curry((args) => {
  history.push({ pathname: '/template_syncs/result' })
})

const submit = (values, dispatch, props) => {
  const { submitForm, syncType, importUrl, exportUrl, history } = props;
  console.log('submitting form, props:');
  console.log(props);
  const url = syncType === 'import' ? importUrl : exportUrl;
  return submitForm({ url, values, message: `Templates were ${syncType}ed.`, item: 'TemplateSync' }).then(
    (args) => {
      console.log('redirect args')
      console.log(args)
      history.replace({ pathname: '/template_syncs/result' })
    }
  )
}

const redirectToResult = (history) => () => history.push({ pathname: '/template_syncs/result' })

class TemplateSyncForm extends React.Component {

  constructor(props) {
    super(props);
    console.log(props)
  }

  radioButtons(syncType) {
    const changeSyncType = (event) => {
      console.log(`Changing sync type to ${event.target.value}`);
      // this.setState({ syncType: event.target.value });
    }

    const checked = (value, syncType) => value === syncType;

    return [
      { label: 'Import', checked: checked("import", syncType), value: "import" },
      { label: 'Export', checked: checked("export", syncType), value: "export" }
    ]
  }

  render() {
    const { submitting,
            error,
            handleSubmit,
            importSettings,
            exportSettings,
            syncType,
            dispatch,
            history,
            validationData,
            valid } = this.props;

    const resetToDefault = curry((dispatch, change, formName, fieldName, value) => {
      dispatch(change(formName, fieldName, value));
    })(dispatch, change, formName);

    return(
      <div>
        <Title titleText="Import or Export Templates" />
        <Form onSubmit={handleSubmit(submit)} disabled={submitting || !valid} submitting={submitting} error={error} onCancel={redirectToResult(history)}>
          <RadioButtonGroup name="syncType" controlLabel="Action type" radios={this.radioButtons(syncType)} disabled={submitting}></RadioButtonGroup>
          <SyncSettingsFields importSettings={importSettings}
                              exportSettings={exportSettings}
                              syncType={syncType}
                              resetField={resetToDefault}
                              disabled={submitting}
                              validationData={validationData}>
          </SyncSettingsFields>
        </Form>
      </div>
    );
  }
}

const prepareInitialValues = (importSettings, exportSettings) =>
  (!importSettings || !exportSettings)
    ? ({})
    : reduce((memo, item) => Object.assign(memo, { [item.name]: item.value }),
             {},
             concat(importSettings, exportSettings));

const mapStateToProps = (state, ownProps) => {
  const initSyncType = { syncType: "import" };
  const syncType = formValueSelector(formName)(state, 'syncType');
  const initialValues = prepareInitialValues(ownProps.importSettings, ownProps.exportSettings);

  if ((!ownProps.importSettings && !ownProps.exportSettings && !syncType) || (ownProps.importSettings && ownProps.exportSettings && syncType)) {
    return Object.assign({ syncType }, { initialValues });
  } else {
    return Object.assign(initSyncType, { initialValues });
  }
}

const form = reduxForm({ form: formName })(TemplateSyncForm);
export default connect(mapStateToProps, FormActions)(form);
