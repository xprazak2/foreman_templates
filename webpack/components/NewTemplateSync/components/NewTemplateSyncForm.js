import React from 'react';
import { reduxForm, formValueSelector, change } from 'redux-form';
import { connect } from 'react-redux';

import Form from 'foremanReact/components/common/forms/Form';
import RadioButtonGroup from 'foremanReact/components/common/forms/RadioButtonGroup';
import * as FormActions from 'foremanReact/redux/actions/common/forms';
import TextField from 'foremanReact/components/common/forms/TextField';

import SyncSettingsFields from './SyncSettingFields';
import Title from '../../layout/Title';

const formName = 'newTemplateSync';

const submit = (values, dispatch, props) => {
  const { submitForm, syncType, importUrl, exportUrl, history } = props;
  const url = syncType === 'import' ? importUrl : exportUrl;
  return submitForm({ url, values, message: `Templates were ${syncType}ed.`, item: 'TemplateSync' }).then(
    (args) => {
      history.replace({ pathname: '/template_syncs/result' })
    }
  )
}

const redirectToResult = (history) => () => history.push({ pathname: '/template_syncs/result' })

const radioButtons = (syncType) => (
  [
    { label: 'Import', checked: ("import" === syncType), value: "import" },
    { label: 'Export', checked: ("export" === syncType), value: "export" }
  ]
);

const TemplateSyncForm = ({
  submitting,
  error,
  handleSubmit,
  importSettings,
  exportSettings,
  syncType,
  dispatch,
  history,
  validationData,
  valid
}) => {

  const resetToDefault = ((dispatch, change, formName) => (fieldName, value) => {
    dispatch(change(formName, fieldName, value));
  })(dispatch, change, formName);

  return (
    <div>
      <Title titleText="Import or Export Templates" />
      <Form onSubmit={handleSubmit(submit)} disabled={submitting || !valid} submitting={submitting} error={error} onCancel={redirectToResult(history)}>
        <RadioButtonGroup name="syncType" controlLabel="Action type" radios={radioButtons(syncType)} disabled={submitting}></RadioButtonGroup>
        <SyncSettingsFields importSettings={importSettings}
                            exportSettings={exportSettings}
                            syncType={syncType}
                            resetField={resetToDefault}
                            disabled={submitting}
                            validationData={validationData}>
        </SyncSettingsFields>
      </Form>
    </div>
  );
}

const prepareInitialValues = (importSettings, exportSettings) =>
  importSettings.concat(exportSettings)
                .reduce((memo, item) => Object.assign(memo, { [item.name]: item.value }), {});

const mapStateToProps = (state, ownProps) => {
  const initSyncType = { syncType: "import" };
  const syncType = formValueSelector(formName)(state, 'syncType');

  const initialValues = prepareInitialValues(ownProps.importSettings, ownProps.exportSettings);

  if (syncType) {
    return ({ initialValues: { ...initialValues, syncType }, syncType });
  } else {
    return ({ initialValues: { ...initialValues, ...initSyncType }, ...initSyncType });
  }
}

const form = reduxForm({ form: formName })(TemplateSyncForm);
export default connect(mapStateToProps, FormActions)(form);
