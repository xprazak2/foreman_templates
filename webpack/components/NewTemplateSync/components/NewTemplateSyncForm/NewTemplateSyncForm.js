import React from 'react';
import PropTypes from 'prop-types';
import { get } from 'lodash';

import Form from 'foremanReact/components/common/forms/Form';
import ForemanForm from 'foremanReact/components/common/forms/ForemanForm';
import { Formik } from 'formik';
import * as Yup from 'yup';
import SyncSettingsFields from '../SyncSettingFields';
import SyncTypeRadios from '../SyncTypeRadios';
// import { formName } from './NewTemplateSyncFormConstants';

const redirectToResult = history => () =>
  history.push({ pathname: '/template_syncs/result' });

const repoFormat = formatAry => value => {
  if (value === undefined) {
    return true;
  }

  const valid = formatAry
    .map(item => (value).startsWith(item))
    .reduce((memo, item) => item || memo, false);

  return value && valid;
};

const syncFormSchema = (syncType, settingsObj, validationData) => {
  const schema = (settingsObj[syncType].asMutable() || []).reduce((memo, setting) => {
    if (setting.name === 'repo') {
      return {
        ...memo,
        repo: Yup.string().test(
          'repo-format',
          `Invalid repo format, must start with one of: ${validationData.repo.join(', ')}`,
          repoFormat(validationData.repo))
        .required("can't be blank")
      }
    }
    return memo
  }, {});

  return Yup.object().shape({
    [syncType]: Yup.object().shape(schema)
  });
}

class NewTemplateSyncForm extends React.Component {
  allowedSyncType = (userPermissions, radioAttrs) =>
    this.props.userPermissions[radioAttrs.permission];

  constructor(props) {
    super(props);

    this.radioButtons = [
      { label: 'Import', value: 'import', permission: 'import' },
      { label: 'Export', value: 'export', permission: 'export' },
    ];

    this.state = {
      syncType: this.radioButtons.find(radioAttrs =>
        this.allowedSyncType(props.userPermissions, radioAttrs)
      ).value,
    };
  }

  updateSyncType = event => {
    this.setState({ syncType: event.target.value });
  };

  permitRadioButtons = buttons =>
    buttons.filter(buttonAttrs =>
      this.allowedSyncType(this.props.userPermissions, buttonAttrs)
    );

  initRadioButtons = syncType =>
    this.permitRadioButtons(this.radioButtons).map(buttonAttrs => ({
      get checked() {
        return this.value === syncType;
      },
      onChange: this.updateSyncType,
      ...buttonAttrs,
    }));

  render() {
    const {
      submitting,
      error,
      submitForm,
      importSettings,
      exportSettings,
      history,
      validationData,
      valid,
      importUrl,
      exportUrl,
      initialValues,
    } = this.props;

    const resetToDefault = (fieldName, fieldValue) => resetFn => {
      return resetFn(fieldName, fieldValue);
    };

    return (
      <ForemanForm
        onSubmit={(values, actions) => {
          const url = this.state.syncType === 'import' ? importUrl : exportUrl;
          return submitForm({
            url,
            values: values[this.state.syncType],
            message: `Templates were ${this.state.syncType}ed.`,
            item: 'TemplateSync',
          })
          .then(args => {
            history.replace({ pathname: '/template_syncs/result' });
          })
        }}
        initialValues={initialValues}
        validationSchema={syncFormSchema(this.state.syncType, { import: importSettings, export: exportSettings }, validationData)}
        onCancel={redirectToResult(history)}
        error={error}
      >
        <SyncTypeRadios
          name="syncType"
          controlLabel="Action type"
          radios={this.initRadioButtons(this.state.syncType)}
        />
        <SyncSettingsFields
          importSettings={importSettings}
          exportSettings={exportSettings}
          syncType={this.state.syncType}
          resetField={resetToDefault}
         />
      </ForemanForm>
    );
  }
}

NewTemplateSyncForm.propTypes = {
  importSettings: PropTypes.array,
  exportSettings: PropTypes.array,
  userPermissions: PropTypes.object.isRequired,
  submitting: PropTypes.bool,
  error: PropTypes.object,
  handleSubmit: PropTypes.func.isRequired,
  dispatch: PropTypes.func,
  history: PropTypes.object,
  validationData: PropTypes.object,
  valid: PropTypes.bool.isRequired,
};

NewTemplateSyncForm.defaultProps = {
  importSettings: [],
  exportSettings: [],
  validationData: {},
  error: undefined,
  dispatch: () => {},
  submitting: false,
  history: {},
};

export default NewTemplateSyncForm;
