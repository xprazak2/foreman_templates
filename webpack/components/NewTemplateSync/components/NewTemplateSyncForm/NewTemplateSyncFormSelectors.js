import { createSelector } from 'reselect';

import {
  selectImportSettings,
  selectExportSettings,
} from '../../NewTemplateSyncSelectors';

const transformInitialValues = (settings) =>
  settings.reduce(
    (memo, item) => Object.assign(memo, { [item.name]: item.value }),
    {}
  )


export const selectInitialFormValues = createSelector(
  selectImportSettings,
  selectExportSettings,
  (importSettings, exportSettings) => {
    return ({ import: transformInitialValues(importSettings),
             export: transformInitialValues(exportSettings)})
  }
);
