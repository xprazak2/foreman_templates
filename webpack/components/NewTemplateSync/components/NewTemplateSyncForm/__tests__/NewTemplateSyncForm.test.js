import { testComponentSnapshotsWithFixtures } from 'react-redux-test-utils';
import Immutable from 'seamless-immutable';

import NewTemplateSyncForm from '../NewTemplateSyncForm';

import {
  importSettings,
  exportSettings,
  initialValues,
} from '../../../__fixtures__/templateSyncSettings.fixtures';

const noop = () => {};

const commonFixture = {
  importUrl: '/import',
  exportUrl: '/export',
  validationData: {
    repo: ['http://', 'ssh://'],
  },
  userPermissions: {
    import: true,
    export: true,
  },
  handleSubmit: noop,
  valid: true,
  importSettings: Immutable(importSettings),
  exportSettings: Immutable(exportSettings),
  initialValues,
  submitForm: () => {},
};

const fixtures = {
  'should render when for import settings': {
    loadingSettings: false,
    ...commonFixture,
  },
  'should render for export settings': {
    loadingSettings: true,
    ...commonFixture,
    userPermissions: {
      import: false,
      export: true,
    },
  },
};

describe('NewTemplateSyncForm', () =>
  testComponentSnapshotsWithFixtures(NewTemplateSyncForm, fixtures));
