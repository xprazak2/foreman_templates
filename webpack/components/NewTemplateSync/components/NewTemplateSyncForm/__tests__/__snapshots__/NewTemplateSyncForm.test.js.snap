// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NewTemplateSyncForm should render for export settings 1`] = `
<ForemanForm
  initialValues={
    Object {
      "export": Object {
        "filter": "",
        "negate": false,
        "repo": "https://github.com/theforeman/community-templates.git",
      },
      "import": Object {
        "associate": "new",
        "force": false,
      },
    }
  }
  onCancel={[Function]}
  onSubmit={[Function]}
  validationSchema={
    ObjectSchema {
      "_blacklist": RefSet {
        "list": Array [],
        "refs": Array [],
      },
      "_conditions": Array [],
      "_defaultDefault": [Function],
      "_deps": Array [],
      "_excludedEdges": Array [],
      "_exclusive": Object {},
      "_mutate": undefined,
      "_nodes": Array [
        "export",
      ],
      "_options": Object {
        "abortEarly": true,
        "recursive": true,
      },
      "_type": "object",
      "_typeError": [Function],
      "_whitelist": RefSet {
        "list": Array [],
        "refs": Array [],
      },
      "fields": Object {
        "export": ObjectSchema {
          "_blacklist": RefSet {
            "list": Array [],
            "refs": Array [],
          },
          "_conditions": Array [],
          "_defaultDefault": [Function],
          "_deps": Array [],
          "_excludedEdges": Array [],
          "_exclusive": Object {},
          "_mutate": undefined,
          "_nodes": Array [
            "repo",
          ],
          "_options": Object {
            "abortEarly": true,
            "recursive": true,
          },
          "_type": "object",
          "_typeError": [Function],
          "_whitelist": RefSet {
            "list": Array [],
            "refs": Array [],
          },
          "fields": Object {
            "repo": StringSchema {
              "_blacklist": RefSet {
                "list": Array [],
                "refs": Array [],
              },
              "_conditions": Array [],
              "_deps": Array [],
              "_exclusive": Object {
                "repo-format": false,
                "required": true,
              },
              "_mutate": undefined,
              "_options": Object {
                "abortEarly": true,
                "recursive": true,
              },
              "_type": "string",
              "_typeError": [Function],
              "_whitelist": RefSet {
                "list": Array [],
                "refs": Array [],
              },
              "tests": Array [
                [Function],
                [Function],
              ],
              "transforms": Array [
                [Function],
              ],
            },
          },
          "tests": Array [],
          "transforms": Array [
            [Function],
          ],
        },
      },
      "tests": Array [],
      "transforms": Array [
        [Function],
      ],
    }
  }
>
  <SyncTypeRadios
    controlLabel="Action type"
    disabled={false}
    name="syncType"
    radios={
      Array [
        Object {
          "checked": true,
          "label": "Export",
          "onChange": [Function],
          "permission": "export",
          "value": "export",
        },
      ]
    }
  />
  <SyncSettingsFields
    exportSettings={
      Array [
        Object {
          "description": "Import or export names matching this regex (case-insensitive; snippets are not filtered)",
          "id": 47,
          "name": "filter",
          "settingsType": "string",
          "value": "",
        },
        Object {
          "description": "Negate the filter for import/export",
          "id": 48,
          "name": "negate",
          "settingsType": "bool",
          "value": false,
        },
        Object {
          "description": "Target path to import and export. Different protocols can be used, for example /tmp/dir, git://example.com, https://example.com, ssh://example.com. When exporting to /tmp, note that production deployments may be configured to use private tmp.",
          "id": 49,
          "name": "repo",
          "settingsType": "string",
          "value": "https://github.com/theforeman/community-templates.git",
        },
      ]
    }
    formProps={Object {}}
    importSettings={
      Array [
        Object {
          "description": "Associate templates to OS, organization and location",
          "id": 45,
          "name": "associate",
          "selection": Array [
            Object {
              "label": "New",
              "value": "new",
            },
            Object {
              "label": "Never",
              "value": "never",
            },
            Object {
              "label": "Always",
              "value": "always",
            },
          ],
          "settingsType": "string",
          "value": "new",
        },
        Object {
          "description": "Should importing overwrite locked templates?",
          "id": 46,
          "name": "force",
          "settingsType": "bool",
          "value": false,
        },
      ]
    }
    resetField={[Function]}
    syncType="export"
  />
</ForemanForm>
`;

exports[`NewTemplateSyncForm should render when for import settings 1`] = `
<ForemanForm
  initialValues={
    Object {
      "export": Object {
        "filter": "",
        "negate": false,
        "repo": "https://github.com/theforeman/community-templates.git",
      },
      "import": Object {
        "associate": "new",
        "force": false,
      },
    }
  }
  onCancel={[Function]}
  onSubmit={[Function]}
  validationSchema={
    ObjectSchema {
      "_blacklist": RefSet {
        "list": Array [],
        "refs": Array [],
      },
      "_conditions": Array [],
      "_defaultDefault": [Function],
      "_deps": Array [],
      "_excludedEdges": Array [],
      "_exclusive": Object {},
      "_mutate": undefined,
      "_nodes": Array [
        "import",
      ],
      "_options": Object {
        "abortEarly": true,
        "recursive": true,
      },
      "_type": "object",
      "_typeError": [Function],
      "_whitelist": RefSet {
        "list": Array [],
        "refs": Array [],
      },
      "fields": Object {
        "import": ObjectSchema {
          "_blacklist": RefSet {
            "list": Array [],
            "refs": Array [],
          },
          "_conditions": Array [],
          "_defaultDefault": [Function],
          "_deps": Array [],
          "_excludedEdges": Array [],
          "_exclusive": Object {},
          "_mutate": undefined,
          "_nodes": Array [],
          "_options": Object {
            "abortEarly": true,
            "recursive": true,
          },
          "_type": "object",
          "_typeError": [Function],
          "_whitelist": RefSet {
            "list": Array [],
            "refs": Array [],
          },
          "fields": Object {},
          "tests": Array [],
          "transforms": Array [
            [Function],
          ],
        },
      },
      "tests": Array [],
      "transforms": Array [
        [Function],
      ],
    }
  }
>
  <SyncTypeRadios
    controlLabel="Action type"
    disabled={false}
    name="syncType"
    radios={
      Array [
        Object {
          "checked": true,
          "label": "Import",
          "onChange": [Function],
          "permission": "import",
          "value": "import",
        },
        Object {
          "checked": false,
          "label": "Export",
          "onChange": [Function],
          "permission": "export",
          "value": "export",
        },
      ]
    }
  />
  <SyncSettingsFields
    exportSettings={
      Array [
        Object {
          "description": "Import or export names matching this regex (case-insensitive; snippets are not filtered)",
          "id": 47,
          "name": "filter",
          "settingsType": "string",
          "value": "",
        },
        Object {
          "description": "Negate the filter for import/export",
          "id": 48,
          "name": "negate",
          "settingsType": "bool",
          "value": false,
        },
        Object {
          "description": "Target path to import and export. Different protocols can be used, for example /tmp/dir, git://example.com, https://example.com, ssh://example.com. When exporting to /tmp, note that production deployments may be configured to use private tmp.",
          "id": 49,
          "name": "repo",
          "settingsType": "string",
          "value": "https://github.com/theforeman/community-templates.git",
        },
      ]
    }
    formProps={Object {}}
    importSettings={
      Array [
        Object {
          "description": "Associate templates to OS, organization and location",
          "id": 45,
          "name": "associate",
          "selection": Array [
            Object {
              "label": "New",
              "value": "new",
            },
            Object {
              "label": "Never",
              "value": "never",
            },
            Object {
              "label": "Always",
              "value": "always",
            },
          ],
          "settingsType": "string",
          "value": "new",
        },
        Object {
          "description": "Should importing overwrite locked templates?",
          "id": 46,
          "name": "force",
          "settingsType": "bool",
          "value": false,
        },
      ]
    }
    resetField={[Function]}
    syncType="import"
  />
</ForemanForm>
`;
